### PR TITLE
Avoid changing RNG state for `DataLoader(..., shuffle=False)`

### LINKED ISSUE #122697: It seems that iterating over a deterministic dataloader changes the seed state
### 🐛 Describe the bug

By switching **eval_testset = False** to **eval_testset = True** in the codes (i.e., whether iterating over a deterministic dataloader with shuffle=False and fixed transforms), the program with a fixed seed gets different results:
```python
import torch
import numpy as np
import random
import torchvision
import torchvision.transforms as transforms

def set_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=32,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=32,
                                         shuffle=False, num_workers=2)

set_seed(42)
eval_testset = True

for epoch in range(10):
    if epoch >= 2:
        break # for the purpose of saving time

    for i, data in enumerate(trainloader):
        inputs, labels = data
        # train the model ...
        if i == 0:
            print("eval_testset=%s, epoch=%d, iter=%d, loaded labels=%s"%(eval_testset, epoch, i, labels.tolist()))

    if eval_testset:
        with torch.no_grad():
            for data in testloader:
                images, labels = data
                # test the model ...
```
**eval_testset = False:** 
```
eval_testset=False, epoch=0, iter=0, loaded labels=[6, 0, 4, 1, 2, 7, 9, 4, 7, 8, 4, 5, 6, 0, 4, 2, 0, 1, 6, 1, 4, 3, 2, 3, 2, 4, 0, 7, 5, 1, 8, 6]
eval_testset=False, epoch=1, iter=0, loaded labels=[8, 7, 1, 6, 2, 3, 0, 7, 7, 2, 9, 8, 8, 7, 5, 9, 0, 6, 9, 6, 7, 5, 9, 3, 7, 8, 9, 4, 4, 4, 0, 7]
```

**eval_testset = True:**
```
eval_testset=True, epoch=0, iter=0, loaded labels=[6, 0, 4, 1, 2, 7, 9, 4, 7, 8, 4, 5, 6, 0, 4, 2, 0, 1, 6, 1, 4, 3, 2, 3, 2, 4, 0, 7, 5, 1, 8, 6]
eval_testset=True, epoch=1, iter=0, loaded labels=[9, 5, 7, 2, 2, 3, 3, 6, 2, 4, 2, 4, 1, 4, 1, 1, 6, 9, 7, 6, 9, 4, 8, 1, 3, 4, 7, 9, 3, 9, 3, 9]
```
I am wondering how can a deterministic dataloader change the seed state of the codes (i.e., resulting in different outputs).

### Versions

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0            py39h5eee18b_1  
[conda] mkl_fft                   1.3.8            py39h5eee18b_0  
[conda] mkl_random                1.2.4            py39hdb19cb5_0  
[conda] numpy                     1.26.3           py39h5f9d8c6_0  
[conda] numpy-base                1.26.3           py39hb5e798b_0  
[conda] pytorch                   1.13.1          py3.9_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.13.1               py39_cu117    pytorch
[conda] torchvision               0.14.1               py39_cu117    pytorch

cc @andrewkho @gokulavasan @SsnL @VitalyFedyunin @dzhulgakov @mruberry @kurtamohler @ejguan

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125370

Fixes #122697

### CODE REVIEW

**Comment by kurtamohler** [comment]
Closing because of https://github.com/pytorch/pytorch/issues/122697#issuecomment-2089584244
